File: NextTaskServer/app/schemas/task.py
from pydantic import BaseModel, field_validator, Field
from typing import Optional, List
from datetime import datetime
from .user import UserResponse

class TaskBase(BaseModel):
    title: str
    content: Optional[str] = None
    status: Optional[str] = "todo"
    priority: Optional[str] = "medium"
    due_date: Optional[datetime] = None

    @field_validator('due_date', mode='before')
    @classmethod
    def validate_due_date(cls, v):
        if v is None:
            return None
        if isinstance(v, str):
            try:
                if len(v) == 10 and v.count('-') == 2:  # YYYY-MM-DD
                    year, month, day = v.split('-')
                    year_int = int(year)
                    if year_int < 1900 or year_int > 2100:
                        raise ValueError(f"Year {year_int} is out of valid range (1900-2100)")
                return datetime.fromisoformat(v)
            except ValueError as e:
                raise ValueError(f"Invalid date format: {v}. Expected format: YYYY-MM-DD or YYYY-MM-DDTHH:MM:SS")
        return v

    @field_validator('status', mode='before')
    @classmethod
    def normalize_status(cls, v):
        if v is None:
            return "todo"
        mapping = {
            'progress': 'in_progress',
            'in progress': 'in_progress',
            'in-progress': 'in_progress'
        }
        s = str(v).strip().lower()
        return mapping.get(s, s)

class TaskCreate(TaskBase):
    workspace_id: int
    assignees_ids: Optional[List[int]] = None

class TaskUpdate(BaseModel):
    title: Optional[str] = None
    content: Optional[str] = None
    status: Optional[str] = None
    priority: Optional[str] = None
    due_date: Optional[datetime] = None
    time_spent: Optional[int] = None
    assignees_ids: Optional[List[int]] = None

# Response schemas
class TaskAssigneeOut(BaseModel):
    id: int
    name: Optional[str] = None
    email: str
    position: Optional[str] = None
    avatar: Optional[str] = None

class AuthorOut(BaseModel):
    id: int
    name: Optional[str] = None
    email: Optional[str] = None
    position: Optional[str] = None
    avatar: Optional[str] = None

class TaskResponse(BaseModel):
    id: int
    title: str
    content: Optional[str] = None
    status: str
    priority: str
    due_date: Optional[datetime] = None
    time_spent: Optional[int] = 0
    created_at: Optional[datetime] = None
    assignees_ids: Optional[List[int]] = None
    assignees: Optional[List[TaskAssigneeOut]] = Field(default_factory=list)
    owner_id: int
    workspace_id: int
    author: Optional[AuthorOut] = None

class TasksResponse(BaseModel):
    tasks: List[TaskResponse]

class TaskAssigneesRequest(BaseModel):
    assignees_ids: List[int]

File: NextTaskClient/src/shared/types/task.ts
// Типы задач. Используются по всему фронтенду
// Внимание: комментарии на русском по требованию стиля проекта

export interface TaskAssignee {
	id: number;
	name: string | null;
	email: string;
	position?: string | null;
	avatar?: string | null;
}

export interface Task {
	id: number;
	title: string;
	content?: string;
	status: string; // 'todo' | 'progress' | 'done'
	priority: string; // 'low' | 'medium' | 'high'
	due_date?: string; // ISO дата
	time_spent?: number; // Потраченное время в секундах
	created_at?: string; // ISO дата создания
	assignees_ids?: number[]; // Массив ID исполнителей
	assignees?: TaskAssignee[] | null; // Полные данные об исполнителях (если вернул API)
	owner_id: number;
	workspace_id: number;
	author?: {
		id: number;
		name: string;
		position?: string | null;
		email?: string;
	};
}

export interface CreateTaskRequest {
	title: string;
	content?: string;
	due_date?: string;
	status: "todo" | "progress" | "done";
	priority: "low" | "medium" | "high";
	assignees_ids?: number[]; // Исполнители (по желанию)
}

export interface UpdateTaskRequest {
	title?: string;
	content?: string;
	due_date?: string;
	status?: "todo" | "progress" | "done";
	priority?: "low" | "medium" | "high";
	time_spent?: number;
	assignees_ids?: number[] | null;
}

export interface TaskResponse {
	task: Task;
	message?: string;
}

export interface TasksResponse {
	tasks: Task[];
}

export interface TaskTimeTrack {
	id: number;
	task_id: number;
	user_id: number;
	time_spent: number;
	comment?: string | null;
	created_at: string;
	user?: {
		id: number;
		name: string | null;
		email: string;
		avatar?: string | null;
		position?: string | null;
	} | null;
}

export interface CreateTimeTrackRequest {
	time_spent: number;
	comment?: string | null;
}

export interface TaskPlan {
	id: number;
	task_id: number;
	user_id: number;
	workspace_id: number;
	date: string; // YYYY-MM-DD
	hours: number;
	created_at: string;
	updated_at?: string;
	created_by: {
		id: number;
		name: string | null;
		email: string;
		avatar?: string | null;
		position?: string | null;
	} | null;
}

export interface CreateTaskPlanRequest {
	task_id: number;
	user_id: number;
	date: string;
	hours: number;
}

export interface UpdateTaskPlanRequest {
	hours: number;
}

export interface PlanningUser {
	id: number;
	email: string;
	name?: string | null;
	position?: string | null;
	avatar?: string | null;
	is_active: boolean;
	created_at: string;
	updated_at?: string | null;
}

export interface PlanningTaskOption {
	id: number;
	title: string;
	workspace_id: number;
}
